lib/semant.ml

```diff
@@ -224,39 +226,32 @@ let check (functions) =
         in
         if idx >= len then failwith "Index out of range."
         else (arr_ele_typ, SSubscript(arr_sexpr, idx_sexpr))
+      | TypeCast(ty, expr) ->
+        let (ty_exp, var) = check_expr symbols expr in
+         (
+          let type_cast_err e1 e2 = 
+            raise (Failure("Cast type not supported from " ^ 
+                          fmt_typ e1 ^ " to " ^ 
+                          fmt_typ e2)) in
+
+           match ty with
+              | IntType -> (
```

Comment:
This pattern matching could be more concise with the help of OCaml's pattern matching guard: https://ocaml.org/manual/patterns.html#sss:pat-variable
```
# let pair_equal = function
    | x, y when x = y -> true
    | _ -> false;;
```
so that we don't need to declare the `type_cast_err` variable.

Reply:
type_cast_err is just a formatted error, which I defined for ease of use.

cuz there are many cases where we need to raise an error

Reply:
might misunderstand my meaning. I found the guard is a pretty useful syntax sugar, for example, the following code could be rewritten
```
  | StrType ->
    if ty_exp = IntType || ty_exp = FloatType || ty_exp = BoolType || ty_exp = StrType 
    then (ty ,STypeCast(ty, (ty_exp, var)))
    else type_cast_err ty_exp ty
```
to 
```
| StrType when IntType || ty_exp = FloatType || ty_exp = BoolType || ty_exp = StrType -> (ty ,STypeCast(ty, (ty_exp, var)))
```
so that we just need to raise the error once :)